channels/offline-channel-js/src/Sender.ts

```diff
@@ -0,0 +1,717 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import dynamicProto from "@microsoft/dynamicproto-js";
+import {
+    BreezeChannelIdentifier, DEFAULT_BREEZE_ENDPOINT, DEFAULT_BREEZE_PATH, DisabledPropertyName, IConfig, utlSetStoragePrefix
+} from "@microsoft/applicationinsights-common";
+import {
+    IAppInsightsCore, IConfiguration, IDiagnosticLogger, IPayloadData, IProcessTelemetryContext, IProcessTelemetryUnloadContext,
+    ITelemetryUnloadState, IUnloadHookContainer, IXHROverride, OnCompleteCallback, SendPOSTFunction, SendRequestReason, TransportType,
+    _eInternalMessageId, _throwInternal, arrForEach, createProcessTelemetryContext, dumpObj, eLoggingSeverity, getExceptionName, getJSON,
+    getLocation, getNavigator, getWindow, isArray, isBeaconsSupported, isFetchSupported, isXhrSupported, objKeys, onConfigChange,
+    useXDomainRequest
+} from "@microsoft/applicationinsights-core-js";
+import { IPromise, createPromise, doAwaitResponse } from "@nevware21/ts-async";
+import { isFunction, isNumber } from "@nevware21/ts-utils";
+import { ILocalStorageConfiguration, IOfflineSenderConfig } from "./Interfaces/IOfflineProvider";
+import { IBackendResponse, XDomainRequest as IXDomainRequest } from "./Interfaces/ISender";
+
+const DefaultOfflineIdentifier = "OfflineChannel";
+//const FetchSyncRequestSizeLimitBytes = 65000; // approx 64kb (the current Edge, Firefox and Chrome max limit)
+const PostChannelId = "PostChannel";
+
+
+declare var XDomainRequest: {
+    prototype: IXDomainRequest;
+    new(): IXDomainRequest;
+};
+
+
+function _getResponseText(xhr: XMLHttpRequest | IXDomainRequest) {
+    try {
+        return xhr.responseText;
+    } catch (e) {
+        // Best effort, as XHR may throw while XDR wont so just ignore
+    }
+
+    return null;
+}
+
+function isOverrideFn(httpXHROverride: any) {
+    return httpXHROverride && httpXHROverride.sendPOST;
+}
+
+function _prependTransports(theTransports: TransportType[], newTransports: TransportType | TransportType[]) {
+    if (newTransports) {
+        if (isNumber(newTransports)) {
+            theTransports = [newTransports as TransportType].concat(theTransports);
+        } else if (isArray(newTransports)) {
+            theTransports = newTransports.concat(theTransports);
+        }
+    }
+    return theTransports;
+}
+
+
+export type SenderFunction = (payload: string[], isAsync: boolean) => void | IPromise<boolean>;
+
+export class Sender {
+
+    public _appId: string; //TODO: set id
+
+    constructor(endpointUrl?: string) {
```

Comment:
Add a TODO: to remove the passing (and internal usages) of the endpointURL, as your already always creating without an endpoint anyway.

If there is no other endpoint (the URL) lets just not send anything

Reply:
Actually, lets add a TODO to remove this class completely so we can delegate the "sending" to the real channels via a new "sendbatch" (or something like that) on IOfflineSupport, that way we don't need to worry about all of the transports etc)

Reply:
Endpoint is removed, and if no url is provided, `doOncomplete` will be called with status code 400 and will throw internal as well.
created Task 26681160 for this to add more tests